TDL11-Terminate_Instances_with_snapshots/terminate_aged_instances_with_snapshots.py

```diff
@@ -0,0 +1,224 @@
+        delete_on_termination = device_type.get('DeleteOnTermination')
+        if delete_on_termination == False:
+            print(f"EBS {device_name} is not set to delete on termination, set it true")
+            response = ec2.modify_instance_attribute(
+                Attribute='blockDeviceMapping',
+                BlockDeviceMappings=[
+                    {
+                        'DeviceName': device_name,
+                        'Ebs': {
+                            'DeleteOnTermination': True,
+                            }
+                    }
+                ],
+                InstanceId=instance_id)
+            print(response)
+        else:
+            print(f"Instance{instance_id}- EBS {device_name} had been set DeleteOnTermination true.")
+
+def snapshot_volumes(instance_id,intances):
+    snapshot_ids=[]
+    print(f"Processing volumes attached to {instance_id} for snapshot")
+    # Do a dryrun first to verify permissions
+    try:
+        response = ec2.create_snapshots(
+            Description='snapshot_'+instance_id,
+            InstanceSpecification={
+                'InstanceId': instance_id,
+                'ExcludeBootVolume': True
+            },
+            DryRun=True,
+            CopyTagsFromSource='volume'
+        )
+    except ClientError as e:
+        if 'DryRunOperation' not in str(e):
+            raise e
+    # Dry run succeeded, call terminate_instances without dryrun
+    try:
+        response = ec2.create_snapshots(
+            Description='snapshot_'+instance_id,
+            InstanceSpecification={
+                'InstanceId': instance_id,
+                'ExcludeBootVolume': True
+            },
+            DryRun=False,
+            CopyTagsFromSource='volume'
+        )
+        logger.info("create snapshots for instance %s.", instance_id)
+        print(response)
+        # Check if snapshots are created completely.
+        for snapid in response.get('Snapshots'):
+            print(snapid.get('SnapshotId'))
+            snapshot_ids.append(snapid.get('SnapshotId'))
+            snapshots_wait(snapshot_ids)
+    except ClientError as e:
+        logger.exception("Couldn't create snapshots %s.", instance_id)
+        raise e
+    
+def terminate_instances(instance_id):
+    print(f"Proceding with termination of {instance_id}")
+    # Do a dryrun first to verify permissions
+    try:
+        ec2.terminate_instances(InstanceIds=[instance_id], DryRun=True)
+    except ClientError as e:
+        if 'DryRunOperation' not in str(e):
+            raise e
+    # Dry run succeeded, call terminate_instances without dryrun
+    try:
+        response = ec2.terminate_instances(InstanceIds=[instance_id], DryRun=False)
+        logger.info("Terminating instance %s.", instance_id)
+        print(response)
+    except ClientError as e:
+        logger.exception("Couldn't terminate instance %s.", instance_id)
+        print(e)
+        raise e
+
+def terminate_main():
+    global ec2 
```

Comment:
Try to avoid using global here.
The better way is to pass ec2 as a parameter to other functions to make functions more independent.
For example：

ec2_client = boto3.client("ec2")
...

# Pass ec2_client as parameter
check_ebs(ec2_client, instance_id,instance)

Reply:
Good advice. Thank you. I had modified it today and will update tomorrow.